### PR TITLE
Mode for adjusting the draw panel using a mouse

### PR DESCRIPTION
**Pull request by jakgin**
Feature presentation:
https://github.com/iis-powp-2024/powp_jobs2d_projects/assets/84508832/837b1202-3a17-47e8-9606-5689fbc83b58

### CODE REVIEW

File: src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapter.java [hunk 1]
```diff
@@ -31,6 +33,9 @@ public void operateTo(int x, int y) {
         line.setStartCoordinates(this.startX, this.startY);
         this.setPosition(x, y);
         line.setEndCoordinates(x, y);
+        try {
+            linesRecorder.addLine((ILine) line.clone());
+        } catch (CloneNotSupportedException ignored) {}
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapter.java]: The solution would be better if LinesRecorder dependency could be avoided. Maybe, the decorator approach is possible by providing a 'line' clone getter in LineDriverAdapter.

**Review by kowallus — CHANGES_REQUESTED** [review]
I have tested the solution and it works quite nice. Please consider the following changes:

1. Major - the adjusted panel is applied to already drawn lines only. The next operations on a driver are executed without transformation. E.g., I have run 'Figure Joe 1' script several times:
image and it is always drawn with the same size and position in the middle.

2. Minor - after zooming out there is a loss in precision of the lines and some lines even collapse into a single point, however, I understand that keeping original coordinates might be difficult.

**Comment by jakgin** [comment]
Major and Minor problem from above fixed. 
Short presentation how it works: https://github.com/iis-powp-2024/powp_jobs2d_projects/assets/84508832/b65942f1-c42f-43df-9508-88d8051957ae
By introducing decorator that wraps LineDriverAdapter it's hard to integrate it to every test. It would require wrapping LineDriverAdapter with the new decorator everywhere. It also not possible to just wrap the currentDriver since the driver might not be an instance of LineDriverAdapter but some other decorator perhaps. It would be perfect to wrap the Job2dDriver instead, but the new decorator needs the lineClone() that only LineDriverAdapter has.

File: src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapterRecordingDecorator.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+public class LineDriverAdapterRecordingDecorator implements Job2dDriver {
+    private final LineDriverAdapter lineDriverAdapter;
+
+    public LineDriverAdapterRecordingDecorator(LineDriverAdapter lineDriverAdapter) {
+        this.lineDriverAdapter = lineDriverAdapter;
+    }
+
+    @Override
+    public void setPosition(int i, int i1) {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapterRecordingDecorator.java:21]: Bad param names

File: src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapterRecordingDecorator.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            Point transformedEndPoint = new ShiftTransformation(currentShift.x, currentShift.y).transform(endPoint);
+
+            line.setStartCoordinates(transformedStartPoint.x, transformedStartPoint.y);
+            line.setEndCoordinates(transformedEndPoint.x, transformedEndPoint.y);
+
+            LinesRecorder.getLinesRecorder().addLine(line, unscaledLine);
+
+            LinesTransformationExecutor linesTransformationExecutor = new LinesTransformationExecutor();
+            double currentZoom = WorkspaceTransformationRecorder.getInstance().getZoom();
+            linesTransformationExecutor.execute(new ScaleTransformation(currentZoom));
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapterRecordingDecorator.java:51]: Not OCP. Handling another transformation requires modification here.

File: src/main/java/edu/kis/powp/jobs2d/features/WorkspaceTransformationFeature.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package edu.kis.powp.jobs2d.features;
+
+import edu.kis.powp.appbase.Application;
+import edu.kis.powp.jobs2d.events.SelectWorkspaceTransformationListener;
+
+public class WorkspaceTransformationFeature {
+    private static boolean isTransforming = false;
+
+    private static DrawPanelMouseMoveFeature drawPanelMouseMoveFeature;
+    private static DrawPanelMouseZoomFeature drawPanelMouseZoomFeature;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/features/WorkspaceTransformationFeature.java:10]: This part could also be more OCP.

File: src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapterRecordingDecorator.java (hunk 3 above)
Comment (kowallus) [reply src/main/java/edu/kis/powp/jobs2d/drivers/adapter/LineDriverAdapterRecordingDecorator.java:51]: Creating a new instance of ScaleTransofrmation every time is a waste of resources.

File: src/main/java/edu/kis/powp/jobs2d/transformations/LinesTransformationExecutor.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        DrawerFeature.getDrawerController().clearPanel();
+
+        List<ILine> lines = linesRecorder.getLines();
+        List<ILine> untransformedLines = linesRecorder.getUntransformedLines();
+
+        for (int i = 0; i < lines.size(); i++) {
+            ILine line = lines.get(i);
+            ILine untransformedLine = untransformedLines.get(i);
+
+            if (transformation instanceof ScaleTransformation) {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/transformations/LinesTransformationExecutor.java:28]: Not Object-Oriented way. Avoiding instanceof is probably possible (maybe Strategy or Visitor?)

**Review by kowallus — APPROVED** [review]
In some places, more OCP would be easily possible. 4.5 project final

**Pull request merged**